### PR TITLE
DEP: Deprecate truth testing on empty arrays

### PR DESCRIPTION
**Pull request by hemildesai**
Add a deprecation warning while truth testing on empty arrays.

Solves step 1 of the proposal in https://github.com/numpy/numpy/issues/9583

### CODE REVIEW

File: numpy/core/src/multiarray/number.c [hunk 1]
```diff
@@ -797,6 +797,12 @@ _array_nonzero(PyArrayObject *mp)
         return res;
     }
     else if (n == 0) {
+        if (DEPRECATE("Returning False. "
+                      "However, the truth value of an empty array is ambiguous "
+                      "and will result in an error in the future. "
+                      "Use a.size > 0") < 0) {
+          return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Nit: Indentation is wrong here

File: numpy/core/src/multiarray/number.c [hunk 2]
```diff
@@ -797,6 +797,12 @@ _array_nonzero(PyArrayObject *mp)
         return res;
     }
     else if (n == 0) {
+        if (DEPRECATE("Returning False. "
+                      "However, the truth value of an empty array is ambiguous "
+                      "and will result in an error in the future. "
+                      "Use a.size > 0") < 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: I think this should be reordered to

    The truth value of an empty array is ambiguous.
    Returning `False`, but in future this will result in an error.
    To check for emptiness, use `a.size > 0` instead.

Remember that warnings are not very good at pointing to their origin - so saying what the problem is should come first

**Comment by eric-wieser** [comment]
This needs a remark in the deprecations section of the release notes.

**Comment by eric-wieser** [comment]
Four test failures, ie bits of numpy code that were using this:

```
======================================================================
ERROR: numpy.core.tests.test_numeric.TestAllclose.test_ip_allclose(inf, inf)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1729, in tst_allclose
    assert_(np.allclose(x, y), "%s and %s not close" % (x, y))
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2466, in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2558, in isclose
    cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2533, in within_tol
    result = bool(result)
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
======================================================================
ERROR: numpy.core.tests.test_numeric.TestIsclose.test_equal_nan
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1892, in test_equal_nan
    assert_array_equal(np.isclose(np.nan, np.nan, equal_nan=True), [True])
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2558, in isclose
    cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2533, in within_tol
    result = bool(result)
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
======================================================================
ERROR: numpy.core.tests.test_numeric.TestIsclose.test_ip_all_isclose(inf, inf)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1860, in tst_all_isclose
    assert_(np.all(np.isclose(x, y)), "%s and %s not close" % (x, y))
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2558, in isclose
    cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2533, in within_tol
    result = bool(result)
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
======================================================================
ERROR: numpy.core.tests.test_numeric.TestIsclose.test_ip_isclose_allclose(inf, inf)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1870, in tst_isclose_allclose
    assert_(np.isclose(x, y) == np.allclose(x, y), msg=msg2 % (x, y))
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2558, in isclose
    cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2533, in within_tol
    result = bool(result)
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
======================================================================
ERROR: numpy.core.tests.test_numeric.TestIsclose.test_non_finite_scalar
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1936, in test_non_finite_scalar
    assert_(np.isclose(np.inf, -np.inf) is False)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2558, in isclose
    cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 2533, in within_tol
    result = bool(result)
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
======================================================================
ERROR: numpy.core.tests.test_numeric.TestNonzero.test_count_nonzero_axis
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/tests/test_numeric.py", line 1024, in test_count_nonzero_axis
    m, axis=np.array([[1], [2]]))
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 473, in assertRaises
    callableObj(*args, **kwargs)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/numeric.py", line 428, in count_nonzero
    if axis is None or axis == ():
DeprecationWarning: Returning False. However, the truth value of an empty array is ambiguous and will result in an error in the future. Use a.size > 0
```

**Comment by hemildesai** [comment]
Looking into them now.

**Comment by hemildesai** [comment]
So, it seems there are 6 errors in total. 5 of them arise from [`bool(result)`](https://github.com/numpy/numpy/blob/master/numpy/core/numeric.py#L2533). This happens due to the fact that [`less_eqaul`](https://github.com/numpy/numpy/blob/master/numpy/core/numeric.py#L2531) returns an empty array if both values are of type `inf` or `nan`. I could go ahead and add a change [here](https://github.com/numpy/numpy/blob/master/numpy/core/numeric.py#L2532) that returns `False` if the array is empty, but that seems like a change to be discussed.

**Comment by hemildesai** [comment]
The last error occurs when comparing a multi-dimensional `np.array` with an empty tuple
```python
>>> np.array([[1],[2]]) == ()
array([], shape=(2, 0), dtype=bool)
```

**Comment by hemildesai** [comment]
Based on the issue and the deprecation message, assuming `bool([])` will result in an error in the future, it makes sense to fix these issues. I can go ahead and fix these issues if that is suggested.

**Comment by eric-wieser** [comment]
@hemildesai: I have a fix that avoids `bool` there altogether at #9720.

A quick fix for the `== ()` would be to change that to `len(axis) == 0`

**Comment by hemildesai** [comment]
@eric-wieser https://github.com/numpy/numpy/pull/9718/commits/bbc9204d6a875d4861a3933c875153d2dbe3dac0 incorporates the error message reordering and the indentation nit.

https://github.com/numpy/numpy/pull/9718/commits/141b67926fca379aa4aecc1ed5706ecb3bfb8ebe fixes the empty tuple comparison. Based on https://docs.scipy.org/doc/numpy/reference/generated/numpy.count_nonzero.html, the only types allowed for `axis` are `int` and `tuple`, so I made it this way. But could also use your new suggestion -> i.e. `not isinstance(axis, np.ndarray) and axis == ()`

**Comment by eric-wieser** [comment]
All set for a rebase

**Comment by hemildesai** [comment]
Rebased off master. If this https://github.com/numpy/numpy/pull/9718/commits/f105a384d40960f25ad0505755ccdd6c5afb506f looks good, I can squash the fixup and push again.

File: numpy/core/numeric.py [hunk 3]
```diff
@@ -406,7 +406,7 @@ def count_nonzero(a, axis=None):
     array([2, 3])
 
     """
-    if axis is None or axis == ():
+    if axis is None or (type(axis) == tuple and axis == ()):
```
Comment (mhvk) [inline numpy/core/numeric.py]: nitpick, but should be `type(axis) is tuple`. Though perhaps this should just be `isinstance(axis, tuple) and axis == ()`? (should we preclude, e.g., a `namedtuple`?)

**Comment by mhvk** [comment]
looks good, only one nitpick!

**Comment by eric-wieser** [comment]
@mhvk: That code path is just an optimization anyway, so it's not too big a deal if we miss cases.

It might also be fine to call `axis = normalize_axis_tuple(axis, ndim)` before the optimized case, which ensures that `axis` is a tuple

**Comment by hemildesai** [comment]
@eric-wieser calling `axis = normalize_axis_tuple(axis, ndim)` before that check causes issues when
- `axis = None` 
- axis is out of bounds of the array's `ndim`
- When the input array to the function is an object that does not have the `ndim` property. eg `list`

I think @mhvk's suggestion of `issinstance(axis, tuple) and axis == ()` works best. Shall I go ahead and incorporate that change?

**Comment by eric-wieser** [comment]
Sure, go with @mhvk's. That's at least equivalent to what was there before.

Thinking about it though (outside this PR) - that test is just wrong, and shouldn't be there at all. `axis=()` means "do it on none of the axes", not "do it on all of them"

**Review by eric-wieser — DISMISSED** [review]

**Comment by eric-wieser** [comment]
Still needs a release note, actually

**Comment by hemildesai** [comment]
Going to add the release note under `DEPRECATIONS` in `doc/release/1.14.0-notes.rst`.

How about - `Truth testing on an empty array is deprecated. To check for emptiness, ``np.array([]).size > 0`` should be used.` for the release note?

**Comment by mhvk** [comment]
Release note sounds good except I think one usually wants to check an array is not empty, so maybe:
`Truth testing on an empty array is deprecated. To check an array is not empty, use ``array.size > 0`` `

**Comment by hemildesai** [comment]
Amended the commit with the release note.

File: numpy/core/src/multiarray/number.c [hunk 4]
```diff
@@ -797,6 +797,11 @@ _array_nonzero(PyArrayObject *mp)
         return res;
     }
     else if (n == 0) {
+        if (DEPRECATE("The truth value of an empty array is ambiguous. "
+                      "Returning False, but in future this will result in an error. "
+                      "To check for emptiness, use `a.size > 0` instead.") < 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Maybe this should change to match the version the release note uses - it is a little clearer.

Comment (hemildesai) [reply numpy/core/src/multiarray/number.c]: Should I add `will raise an error in the future` in the first line?

Comment (eric-wieser) [reply numpy/core/src/multiarray/number.c]: I think just change the last line to ``To check an array is not empty, use ``array.size > 0```

**Comment by charris** [comment]
@hemildesai  When you fix the error message, please add a comment before the deprecation
```
/* 2017-24-09, 1.14 */
```
The day doesn't need to be exact, but the message helps keep track of how long deprecations have been out there.

**Comment by eric-wieser** [comment]
@charris, please don't use yyyy-dd-mm as a date format, ever! (I've edited your post to fix that)

**Comment by hemildesai** [comment]
Made the suggested changes and pushed. 👍

File: numpy/core/tests/test_multiarray.py [hunk 5]
```diff
@@ -6541,6 +6541,11 @@ def __bool__(self):
             Error = RuntimeError  # python < 3.5
         assert_raises(Error, bool, self_containing)  # previously stack overflow
 
+        with warnings.catch_warnings(record=True) as w:
+            warnings.filterwarnings('always', '.*truth value of an empty array is ambiguous*.', DeprecationWarning)
+            assert_equal(bool(np.array([])), False)
+            assert_(w[0].category is DeprecationWarning)
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: I think there's an `assert_deprecated` that you ought to be using here (and would be much more concise)

Comment (seberg) [reply numpy/core/tests/test_multiarray.py]: `assert_deprecated` is only defined in `test_deprecations`, `assert_warns` would be fine here too, but if you want to have the warning text, I am not sure you can make it much shorter.

Since this is C-code, I would like a test for the error path as well though.

Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: In that case, this test should probably move to `test_deprecations.py`

**Comment by eric-wieser** [comment]
I've gone ahead and fixed the tests for you. Look good, @seberg?

**Comment by hemildesai** [comment]
Thanks @eric-wieser. Sorry, was a little caught up the past few days. Let me know if there's any other change to be made.

**Comment by seberg** [comment]
Seems good to me, the tuple change in nonzero confuses me a bit, is it here by accident?

**Comment by hemildesai** [comment]
@seberg That change was to fix a regression test failure after adding the deprecation warning. The failure there was
```python
# comparing a multi-dimensional np.array with an empty tuple
>>> np.array([[1],[2]]) == ()
array([], shape=(2, 0), dtype=bool)
```

**Comment by seberg** [comment]
Yes, but where does that come from? From some weirder test? In that case maybe the test should be changed instead of this code? this seems like a slight behaviour change that maybe we should avoid?

**Comment by eric-wieser** [comment]
@seberg: There's a weird special case (#9728) in `count_nonzero` that handles `axis=()`, so it does a test that `axis == ()`.

**Comment by eric-wieser** [comment]
I'm gonna go ahead and put this in. There is something screwy in `count_nonzero`, but it's always been there, is out of scope for this patch, and might even need a `FutureWarning`

**Comment by eric-wieser** [comment]
Welp, this introduces an unintended warning:

```
>>> v0 = np.array(b'', np.void)
>>> bool(v0)
False #ok
>>> v0.astype(bool)
DeprecationWarning: The truth value of an empty array is ambiguous.
```

Of course, that was broken anyway:
```
>>> np.array(b'a', np.void).astype(bool)
array(True, dtype=bool)  # ok
>>> v2 = np.array(b'aa', np.void)
ValueError: setting an array element with a sequence.
```

I think @ahaldane's #8157 will fix this

**Comment by pv** [comment]
More probably unintended warnings in Scipy:
https://github.com/scipy/scipy/pull/8019

**Comment by pv** [comment]
No, sorry the scipy failures look as intended...
